File: SearchDemo/SearchDemo/ViewController.m
// ===================================================================================================
// Copyright © 2018 nanorep.
// NanorepUI SDK.
// All rights reserved.
// ===================================================================================================

#import "ViewController.h"
#import "NanorepUI.h"

@interface ViewController () <NanorepPersonalInfoHandler, UITextFieldDelegate, NRApplicationContentHandler, NRReadMoreViewControllerDelegate>
@property (weak, nonatomic) IBOutlet UITextField *kbTF;
@property (weak, nonatomic) IBOutlet UITextField *accountTF;
@property (weak, nonatomic) IBOutlet UISwitch *phoneConfirmation;
@property (nonatomic, strong) NRWidgetViewController *widgetController;
@property (weak, nonatomic) IBOutlet UITextField *articleIdTF;
@property (weak, nonatomic) IBOutlet UILabel *appVersionLabel;
@property (weak, nonatomic) IBOutlet UILabel *sdkVersion;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    _appVersionLabel.text = [_appVersionLabel.text stringByAppendingString:[[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"]];
    _sdkVersion.text = [_sdkVersion.text stringByAppendingString:[NSBundle sdkVersion]];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (IBAction)loadNanorep:(UIButton *)sender {
    AccountParams *account = [AccountParams new];
    account.account = _accountTF.text;
    account.knowledgeBase = _kbTF.text;
    
//    account.nanorepContext =  @{@"DeviceType": @"iOS"};
    [[NanoRep shared] prepareWithAccountParams:account];
    __weak ViewController *weakSelf = self;
    [NanoRep shared].fetchConfiguration = ^(NRConfiguration *config, NSError *err) {
        if (err) {
            NSLog(@"ERROR ::%@", err);
        } else {
            dispatch_async(dispatch_get_main_queue(), ^{
                weakSelf.widgetController = [NRWidgetViewController new];
                weakSelf.widgetController.infoHandler = weakSelf;
                [weakSelf presentViewController:weakSelf.widgetController animated:YES completion:nil];
            });
        }
    };
}

- (BOOL)shouldOverridePhoneChannel:(NRChannelingPhoneNumber *)phoneChannel {
    if (_phoneConfirmation.isOn) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Phone Confirmation" message:[@"Need your approval for dialing number :\n" stringByAppendingString:phoneChannel.phoneNumber] preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *action = [UIAlertAction actionWithTitle:@"Approve" style:UIAlertActionStyleDefault handler:nil];
        [alert addAction:action];
        [_widgetController presentViewController:alert animated:YES completion:nil];
    }
    return _phoneConfirmation.isOn;
}

- (void)personalInfoWithExtraData:(NSDictionary *)extraData
                          channel:(NRChanneling *)channel
                completionHandler:(void(^)(NSDictionary *formData))handler {
    handler(extraData);
}

- (void)didFetchExtraData:(ExtraData *)formData {
    
}


- (void)didSubmitForm {
    
}

- (void)didCancelForm {
    
}

- (void)presentFileList:(NSArray<NSString *> *)files {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Files To Upload" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    for (NSString *key in files) {
        [alert addAction:[UIAlertAction actionWithTitle:key style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            
        }]];
    }
    [self.widgetController presentViewController:alert animated:YES completion:nil];
}

- (void)didFailSubmitFormWithError:(NSError *)error {
    
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    
    
    // Deep link example
    AccountParams *account = [AccountParams new];
    account.account = _accountTF.text;
    account.knowledgeBase = _kbTF.text;
    [[NanoRep shared] prepareWithAccountParams:account];
    __weak ViewController *weakSelf = self;
    [NanoRep shared].fetchConfiguration = ^(NRConfiguration *config, NSError *err) {
        if (err) {
            NSLog(@"ERROR ::%@", err);
        } else {
            dispatch_async(dispatch_get_main_queue(), ^{
                NRReadMoreViewController *deepLink = [NRReadMoreViewController new];
                deepLink.infoHandler = self;
                deepLink.delegate = self;
                deepLink.articleId = textField.text;
                deepLink.applicationHandler = weakSelf;
                UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:deepLink];
                navController.navigationBar.topItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemStop target:self action:@selector(dismissDeeplinkPage:)];
                [weakSelf presentViewController:navController animated:YES completion:nil];
            });
        }
    };
    return YES;
}

- (void)dismissDeeplinkPage:(UIBarButtonItem *)sender {
    [self.presentedViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)readmoreController:(NRReadMoreViewController *)readmoreController presentModally:(UIViewController *)controller {
    [readmoreController.navigationController presentViewController:controller animated:YES completion:nil];
}

@end

File: SearchDemo/SearchDemo-swift/ViewController.swift
// ===================================================================================================
// Copyright © 2018 nanorep.
// NanorepUI SDK.
// All rights reserved.
// ===================================================================================================

import UIKit
import NanorepUI

class ViewController: UIViewController, NanorepPersonalInfoHandler, UITextFieldDelegate, NRApplicationContentHandler, NRReadMoreViewControllerDelegate  {
    
    /************************************************************/
    // MARK: - Properties
    /************************************************************/
    
    // Outlets
    @IBOutlet weak var kbTF: UITextField!
    @IBOutlet weak var accountTF: UITextField!
    @IBOutlet weak var phoneConfirmation: UISwitch!
    @IBOutlet weak var articleIdTF: UITextField!
    @IBOutlet weak var appVersionLabel: UILabel!
    @IBOutlet weak var sdkVersion: UILabel!
    
    // Params
    var widgetController: NRWidgetViewController!
    
    /************************************************************/
    // MARK: - Functions
    /************************************************************/

    override func viewDidLoad() {
        super.viewDidLoad()
        self.appVersionLabel.text = appVersionLabel.text! + (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "")
        self.sdkVersion.text = sdkVersion.text! + (Bundle.sdkVersion())
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    
    @IBAction func loadNanorep(_ sender: Any) {
        self.view.endEditing(true)
        
        let accountParams = AccountParams()
        accountParams.account =  self.accountTF.text
        accountParams.knowledgeBase = self.kbTF.text
//        accountParams.perform(Selector("setServer:"), with: "qa09")
        NanoRep.shared().prepare(with: accountParams)
        
        NanoRep.shared().fetchConfiguration = {
            (configuration: NRConfiguration?, error: Error?) -> Void in
            
            if error != nil {
                print(error.debugDescription)
            }

            configuration?.useLabels = true
            DispatchQueue.main.async { [unowned self] in
                self.widgetController = NRWidgetViewController()
                self.widgetController.infoHandler = self
                self.present(self.widgetController, animated: true, completion: nil)
//                self.widgetController.view.frame = CGRect(origin: CGPoint.zero, size: CGSize(width: self.view.frame.size.width, height: self.view.frame.size.height - 44.0))
//                self.addChildViewController(self.widgetController)
//                self.view.addSubview(self.widgetController.view)
            }
        }
    }

    /************************************************************/
    // MARK: - NRReadMoreViewControllerDelegate
    /************************************************************/
    
    
    func shouldOverridePhoneChannel(_ phoneChannel: NRChannelingPhoneNumber!) -> Bool {
        if self.phoneConfirmation.isOn {
            let alert = UIAlertController(title: "Phone Confirmation", message: "Need your approval for dialing number :\n" + (phoneChannel?.phoneNumber ?? ""), preferredStyle: .alert)
            let action = UIAlertAction(title: "Approve", style: .default, handler: nil)
            alert.addAction(action)
            widgetController.present(alert, animated: true)
        }
        return phoneConfirmation.isOn
    }
    
//    func shouldPresentConfirmationPopup(forChannel channel: NRChanneling!) -> Bool {
//        return true
//    }

    
    func presentFileList(_ files: [String]?) {
        let alert = UIAlertController(title: "Files To Upload", message: nil, preferredStyle: .actionSheet)
        for key: String? in files ?? [String?]() {
            alert.addAction(UIAlertAction(title: key, style: .default, handler: { action in
            }))
        }
        widgetController.present(alert, animated: true)
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        // Deep link example
        let account = AccountParams()
        account.account = accountTF.text
        account.knowledgeBase = kbTF.text
        NanoRep.shared().prepare(with: account)
        weak var weakSelf: ViewController? = self
        NanoRep.shared().fetchConfiguration = { config, err in
            if (err != nil) {
                print("ERROR ::\(String(describing: err))")
            } else {
                DispatchQueue.main.async(execute: {
                    let deepLink = NRReadMoreViewController()
                    deepLink.infoHandler = self
                    deepLink.articleId = textField.text
                    deepLink.applicationHandler = weakSelf
                    let navController = UINavigationController(rootViewController: deepLink)
                    navController.navigationBar.topItem?.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .stop, target: self, action: #selector(self.dismissDeeplinkPage(_:)))
                    weakSelf?.present(navController, animated: true)
                })
            }
        }
        
        return true
    }
    
    func readmoreController(_ readmoreController: NRReadMoreViewController!, presentModally controller: UIViewController!) {
        if let aController = controller {
            readmoreController?.navigationController?.present(aController, animated: true)
        }
    }
    
    @objc func dismissDeeplinkPage(_ sender: UIBarButtonItem?) {
        ((self.presentedViewController as! UINavigationController).viewControllers.first as! NRReadMoreViewController).dismiss()
        self.presentedViewController?.dismiss(animated: true, completion: {
            
        })
    }
    
    func presenting(_ controller: UIViewController!, shouldHandleClickedLink link: String!) -> Bool {
        return true
    }
    
    func didSubmitFeedback(_ info: [AnyHashable : Any]!) {}
    func didClick(toCall phoneNumber: String!) {}
    func didClickLink(_ url: String!) {}
    func didSessionExpire() {}
    
    func didFetch(_ formData: ExtraData!) {
        print("didFetchExtraData")
        for value in formData.values {
            print(value.value)
        }
    }
    
    func didSubmitForm() {
        print("didSubmitForm")
    }
    
    func didCancelForm() {
        print("didCancelForm")
    }
    
    func personalInfo(withExtraData extraData: [AnyHashable : Any]!, channel: NRChanneling!, completionHandler handler: (([AnyHashable : Any]?) -> Void)!) {
//        var temp = extraData!
//        temp["Testing"] = "working"
        handler(extraData)
        print("personalInfo")
    }
    
    
    func didFailSubmitFormWithError(_ error: Error!) {
        
    }
    func controller(_ controller: NRConversationalViewController!, didClickApplicationQuickOption quickOption: NRQuickOption!) {}
}
